services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/switches/SwitchRulesSpec.groovy

```diff
@@ -13,107 +18,179 @@ import org.openkilda.testing.service.northbound.NorthboundService
+            assert northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.size() == data.rulesRemained
+        }
+
+        and: "Delete the flow"
+        northboundService.deleteFlow(flow.id)
+
+        and: "Install default rules if necessary"
+        if (data.rulesRemained == 0) {
+            northboundService.installSwitchRules(srcSwitch.dpId, InstallRulesAction.INSTALL_DEFAULTS)
+            Wrappers.wait(RULES_INSTALLATION_TIME) {
+                assert northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.size() == defaultRules.size()
+            }
+        }
+
+        where:
+        data << [[description      : "non-default",
+                  deleteRulesAction: DeleteRulesAction.IGNORE_DEFAULTS,
+                  rulesDeleted     : flowRulesCount,
+                  rulesRemained    : defaultRules.size()
+                 ],
+                 [description      : "all",
+                  deleteRulesAction: DeleteRulesAction.DROP_ALL,
+                  rulesDeleted     : defaultRules.size() + flowRulesCount,
+                  rulesRemained    : 0
+                 ]
+        ]
+    }
+
+    def "Able to synchronize rules on a switch (install missing rules)"() {
+        given: "A switch with missing rules"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        northboundService.addFlow(flow)
         Wrappers.wait(WAIT_OFFSET) {
-            northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.size() == defaultRules.size()
+            assert northboundService.getFlowStatus(flow.id).status == FlowState.UP
+            assert northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.size() ==
+                    defaultRules.size() + flowRulesCount
+        }
+
+        def defaultPlusFlowRules = northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.sort { it.cookie }
+        northboundService.deleteSwitchRules(srcSwitch.dpId, DeleteRulesAction.IGNORE_DEFAULTS)
+        Wrappers.wait(RULES_DELETION_TIME) {
+            assert northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.size() == defaultRules.size()
+        }
+        assert northboundService.validateSwitchRules(srcSwitch.dpId).missingRules.size() == flowRulesCount
+
+        when: "Synchronize rules on the switch"
+        def synchronizedRules = northboundService.synchronizeSwitchRules(srcSwitch.dpId)
+
+        then: "The corresponding rules are installed on the switch"
+        synchronizedRules.installedRules.size() == flowRulesCount
+        Wrappers.wait(RULES_INSTALLATION_TIME) {
+            assert expect(northboundService.getSwitchRules(srcSwitch.dpId).flowEntries.sort { it.cookie },
+                    sameBeanAs(defaultPlusFlowRules)
+                            .ignoring("byteCount")
+                            .ignoring("packetCount")
+                            .ignoring("durationNanoSeconds")
+                            .ignoring("durationSeconds"))
```

Comment:
does it make sense to have a local helper method for this comparison? It looks quite bulky and repeats 3 times already

Reply:
Will fix.

Reply:
done.